### PR TITLE
Added handling for Excel Strict ISO dates.

### PR DESCRIPTION
**Pull request by jmcnamara**
Added handling for Excel files with dates in Excel Strict XML format.
This relates to issue #86.

### CODE REVIEW

**Comment by jmcnamara** [comment]
I've rebased this PR to the current master.

File: tests/test_strict_dates.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        Test for dates and times in the Excel standard 1900 epoch.
+
+        """
+        workbook = xlrd.open_workbook(from_this_dir('strict_date_1900.xlsx'))
+        worksheet = workbook.sheet_by_name('Sheet1')
+
+        for row in range(19):
+            num_cell = worksheet.cell(row, 0)
+            date_cell = worksheet.cell(row, 1)
+            self.assertAlmostEqual(num_cell.value, date_cell.value)
```
Comment (cjw296) [inline tests/test_strict_dates.py]: Still need an answer to my question...

File: tests/test_strict_dates.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    def test_strict_date_1904_epoch(self):
+        # Test for dates and times in the Excel for Mac 1904 epoch.
+
+        workbook = xlrd.open_workbook(from_this_dir('strict_date_1904.xlsx'))
+        worksheet = workbook.sheet_by_name('Sheet1')
+
+        for row in range(19):
+            num_cell = worksheet.cell(row, 0)
+            date_cell = worksheet.cell(row, 1)
+            self.assertAlmostEqual(num_cell.value, date_cell.value)
```
Comment (cjw296) [inline tests/test_strict_dates.py]: ...and here too!

File: xlrd/xldate.py [hunk 3]
```diff
@@ -130,6 +131,76 @@ def xldate_as_datetime(xldate, datemode):
     return epoch + datetime.timedelta(days, seconds, 0, milliseconds)
 
 
+##
+# Convert an Excel date/time in ISO date format into a serial number.
+#
+# @param xldate The Excel number
```
Comment (cjw296) [inline xlrd/xldate.py]: Indeed *sigh*

File: xlrd/xldate.py [hunk 4]
```diff
@@ -130,6 +131,76 @@ def xldate_as_datetime(xldate, datemode):
...
+        # Excel for Mac date epoch.
+        epoch = epoch_1904
+    else:
+        # Default Excel epoch.
+        epoch = epoch_1900_minus_1
+
+    if re.search("T", xldate):
+        # The Excel date is in a date and time format.
+        m = re.match(r'(\d\d\d\d)-(\d\d)-(\d\d)T(\d\d):(\d\d):(\d\d)(.\d+)',
+                     xldate)
```
Comment (cjw296) [inline xlrd/xldate.py]: Do we really need a date, though? datetime.strptime shoulld be able to handle this, right?

File: xlrd/xldate.py [hunk 5]
```diff
@@ -130,6 +131,76 @@ def xldate_as_datetime(xldate, datemode):
...
+        raise XLDateError(xldate)
+
+    # Convert a Python datetime.datetime value to an Excel date number.
+    delta = dt - epoch
+    excel_date = (delta.days
+                  + (float(delta.seconds)
+                     + float(delta.microseconds) / 1E6)
+                  / (60 * 60 * 24))
+
+    return excel_date
```
Comment (cjw296) [inline xlrd/xldate.py]: True, but the regex approach feels flawed to me, there must be a better way?

**Comment by cjw296** [comment]
Looks like another rebase is needed at the very least, but surely there must be a way we can do this without a tonne of regex?

File: tests/test_strict_dates.py (hunk 1 above)
Comment (jmcnamara) [inline tests/test_strict_dates.py]: It is to allow for a floating point comparison:

```python
>           self.assertEqual(num_cell.value, date_cell.value)
E           AssertionError: 0.99999 != 0.9999899999999999

``

**Comment by jmcnamara** [comment]
The PR is from 18 months ago and I don't remember why I got involved with it.

I don't have the energy or the interest to rework it so I'm just going to close it. If someone else encounters this issue (fairly unlikely since ISO format dates since aren't the default in Excel) they can use the tests or some of the code to supply a new PR.

**Comment by cjw296** [comment]
The effort is very much appreciated, and the test cases provide a good start.
Thanks for the help, apologies if any of my responses came off wrong...

**Pull request closed without merging**